synapse/lib/node.py

```diff
@@ -196,43 +203,54 @@ def getTag(self, name, defval=None):
         name = s_chop.tag(name)
         return self.tags.get(name, defval)
 
-    def addTag(self, tag, valu=None):
+    def addTag(self, tag, valu=(None, None)):
 
         name = s_chop.tag(tag)
 
-        if not self.xact.allowed('node:tag:add', name):
+        if not self.snap.allowed('node:tag:add', name):
             raise s_exc.AuthDeny()
 
-        parts = name.rsplit('.', 1)
-        if len(parts) > 1:
-            self._addTagRaw(parts[0], None)
+        if valu != (None, None):
+            valu = self.snap.model.type('ival').norm(valu)[0]
 
-        self._addTagRaw(name, valu)
+        curv = self.tags.get(name)
+        if curv == valu:
+            return
 
-    def _addTagRaw(self, name, valu):
+        if curv is not None:
+            # merge tag and move along...
+            return
 
-        node = self.xact.addNode('syn:tag', name)
+        tags = s_chop.tags(name)
+        for tag in tags[:-1]:
 
-        ivaltype = self.xact.model.type('ival')
-        norm, info = ivaltype.norm(valu)
+            if self.tags.get(tag) is not None:
+                continue
 
-        curv = self.tags.get(name, s_common.novalu)
-        if curv == norm:
-            return
+            self._addTagRaw(tag, (None, None))
+
+        self._addTagRaw(tags[-1], valu)
+
+    def _addTagRaw(self, name, norm):
+
+        # these are cached based on norm...
+        self.snap.addTagNode(name)
 
         info = {'univ': True}
-        indx = ivaltype.indx(norm)
+        if norm == (None, None):
+            indx = b'\x00'
+        else:
+            indx = self.snap.model.types['ival'].indx(norm)
+
         sops = (
-            ('prop:set', (self.buid, self.form.name, '#' + name, valu, indx, info)),
+            ('prop:set', (self.buid, self.form.name, '#' + name, norm, indx, info)),
         )
 
         self.tags[name] = norm
-        self.xact.stor(sops)
-
-        if curv is s_common.novalu:
-            #TODO: splice
-            pass
+        self.snap.stor(sops)
 
+        # TODO: fire an onTagAdd handler...
+        self.snap.splice('tag:add', ndef=self.ndef, tag=name, valu=norm)
```

Comment:
tag add splices currently have the ival, if present, set to the `ival` key.  see https://github.com/vertexproject/synapse/blob/v0.0.55/synapse/cores/common.py#L1864

Reply:
For new tags, the ival *is* the valu so, this should be inclusive.